python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -1115,6 +1115,15 @@ class Cast(Expr):
     def __init__(self, dtype: plc.DataType, value: Expr) -> None:
         super().__init__(dtype)
         self.children = (value,)
+        if (
+            self.dtype.id() == plc.TypeId.STRING
+            or value.dtype.id() == plc.TypeId.STRING
+        ):
+            raise NotImplementedError(
+                "Need to implement cast to/from string separately."
+            )
+        elif plc.traits.is_chrono(self.dtype) and plc.traits.is_numeric(value.dtype):
+            raise NotImplementedError("Can't cast duration to numeric")
```

Comment:
I think this is not right. You can cast between duration types and numeric types, but you _can't_ cast between timestamp types and numeric types.

And you can't do so in either direction, so I think this wants to be:

```
elif (plc.traits.is_timestamp(self.dtype) and plc.traits.is_numeric(value.dtype)) or (plc.traits.is_numeric(self.dtype) and plc.traits.is_timestamp(value.dtype):
```

The way, I think, you're supposed to do this in libcudf is (integral -> timestamp) first cast the integral to a duration time, and then cast the duration type to a timestamp type. The converse (timestamp -> integral) does the same in the reverse order.

So we could support this by inserting this double cast in the translation layer, sort of like this (hacky, don't merge, since it hard-codes the offsets between timestamp and duration enums):

```patch
diff --git a/python/cudf_polars/cudf_polars/dsl/translate.py b/python/cudf_polars/cudf_polars/dsl/translate.py
index dec45679c7..419935d2f8 100644
--- a/python/cudf_polars/cudf_polars/dsl/translate.py
+++ b/python/cudf_polars/cudf_polars/dsl/translate.py
@@ -459,6 +459,12 @@ def _(node: pl_expr.Cast, visitor: NodeTraverser, dtype: plc.DataType) -> expr.E
         # Translation of Len/Count-agg put in a cast, remove double
         # casts if we have one.
         (inner,) = inner.children
+    if plc.traits.is_numeric(dtype) and plc.traits.is_timestamp(inner.dtype):
+        intermediate = plc.DataType(plc.TypeId(inner.dtype.id() + 5))
+        return expr.Cast(dtype, expr.Cast(intermediate, inner))
+    elif plc.traits.is_numeric(inner.dtype) and plc.traits.is_timestamp(dtype):
+        intermediate = plc.DataType(plc.TypeId(dtype.id() + 5))
+        return expr.Cast(dtype, expr.Cast(intermediate, inner))
     return expr.Cast(dtype, inner)
```

Reply:
Does this need to be addressed still?

Reply:
We have punted on the casting for now, so it does not.